digitaltwin_client/src/dt_interface_client.c

```diff
@@ -629,11 +700,19 @@ void DT_InterfaceClient_RegistrationCompleteCallback(DIGITALTWIN_INTERFACE_CLIEN
         {
             SetInterfaceState(dtInterfaceClient, DT_INTERFACE_STATE_REGISTERED);
         }
-        (void)InvokeBindingInterfaceUnlock(dtInterfaceClient, &lockHeld);
 
         if (dtInterfaceClient->dtInterfaceRegisteredCallback != NULL)
         {
+            // Mark the fact we're in middle of a callback so that if a call to destroy on this handle arrives at same time,
+            // we'll block destroy call until we exit.  This is basically what BeginInterfaceCallbackProcessing does, 
```

Comment:
>This is basically what BeginInterfaceCallbackProcessing does [](start = 56, length = 60)

Nit: I think following consistent pattern of using our own BeginCallback and EndCallback throughout is more important than a tiny optimization here. A comment helps though if this optimization is really needed.

Reply:
I'd have preferred to route everything through BeginInterfaceCallbackProcessing().  However, Begin...() assumes that the lock is NOT held already.  And the way our locking works, I can't call Lock() on the same lock multiple times a la EnterCriticalSection().

So... rather than having an 
InterfaceUnlock();
BeginInterfaceCallbackProcessing(). // which locks & unlocks
dtInterfaceRegisteredCallback(...);
EndInterfaceClient();

Beyond just optimization, it's safer once we go into the set state DT_INTERFACE_STATE_REGISTERED to do the processingCallback=TRUE under the same lock instead of releasing the lock (which has opening for a destroy to come in before processing when locks are released).

---
In reply to: [324348889](https://github.com/Azure/azure-iot-sdk-c/pull/1173#discussion_r324348889) [](ancestors = 324348889)